app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,474 @@
+# == Schema Information
+#
+# Table name: services
+#
+#  id                    :integer          not null, primary key
+#  type                  :string(255)
+#  title                 :string(255)
+#  project_id            :integer
+#  created_at            :datetime
+#  updated_at            :datetime
+#  active                :boolean          default(FALSE), not null
+#  properties            :text
+#  template              :boolean          default(FALSE)
+#  push_events           :boolean          default(TRUE)
+#  issues_events         :boolean          default(TRUE)
+#  merge_requests_events :boolean          default(TRUE)
+#  tag_push_events       :boolean          default(TRUE)
+#  note_events           :boolean          default(TRUE), not null
+#
+
+require 'digest'
+require 'composer'
+
+class ComposerService < Service
+  prop_accessor :package_mode, :package_type, :export_branches, :branch_filters,
+                :export_tags, :tag_filters, :custom_json
+
+  validates :package_mode, :package_type, presence: true, if: :activated?
+
+  validates_each :custom_json,
+    if: :allow_custom_json_validation? do |record, attr, value|
+      begin
+        name_re = /([A-Za-z0-9&_-]+\/[A-Za-z0-9&_-]+)/
+
+        if value.empty?
+          error = 'must be specified'
+        elsif (cjson = ActiveSupport::JSON.decode(value))
+          if cjson.empty?
+            error = 'must not be empty'
+          elsif !cjson['name']
+            error = 'must have a name key specified'
+          elsif cjson['name'].empty?
+            error = 'name key must not be empty'
+          elsif cjson['name'] != (name_re.match(cjson['name']) || [])[0]
+            error = 'name key must be formatted as "namespace/project"'
+          elsif !cjson['description']
+            error = 'must have a description key specified'
+          elsif cjson['description'].empty?
+            error = 'description key must not be empty'
+          end
+        end
+        record.errors.add(attr, error) if error
+
+      rescue
+        record.errors.add(attr, 'must be a valid JSON string')
+      end
+    end
+
+  after_save :process_project
+
+  def allow_custom_json_validation?
+    activated? && package_mode == 'advanced'
+  end
+
+  def title
+    'Composer'
+  end
+
+  def description
+    'List your project as a composer package'
+  end
+
+  def help
+    out = 'This project will be publicly listed as a composer package,
```

Comment:
Talking from my own case. I'd prefer if this file was behind Basic Auth or similar. It would not be reasonable to disallow this service for all private projects since it's quite likely that's what people would want to use it for, at least I want it for generating a packages.json file out of my private repos, which are used by other private repos :)